Add information for node operators (#94)
vegaprotocol/documentation: docs/node-operators/network_restarts.md

If any differs please mention it on discord.

### Step 0: Edit configuration

You can prepare you configuration but you may not want to update your mainnet node with these before it's stopped.

#### Add the tendermint public key to your nodewallet

The nodewallet now requires to know the tendermint public key. You can save this key by running the following vega command:
```
vega nodewallet --home=/path/to/vega/home import --chain=tendermint --tendermint-pukey=YOUR_TENDERMINT_PUBEY
```

#### Update the vega configuration

New fields have been added to the vega configuration. We recommand you generate a new default config and ensure that no sections are missing from the configuration if the default are not correct for you.

The configuration needs now to specify if a node is running as a validator or not, in the case of a vega validator node (taking part of tendermint consensus) we will want to add the following line to the vega configuration:
```
NodeMode = "validator"
```

#### Update the data-node configuration

As for the vega configuration, new fields have been added in the data-node configuration, we recommend you generate a default one to compare with what you used today.

> The vega node connected to the data-node should add the following line to its config:

Reviewer: Is this adding the same config `NodeMode = "validator"` to the data-node as we have for the core node above?

If so it was not quite clear, I wondered if something like:

''' If the configuration of the core vega node is set to validator, this also needs to be set 
ion the data-node configuration. This is done using the following line: ```

```
NodeMode = "validator"
```

> The data-node configuration also support ssl for the http connection. We recommend that you setup ssl on your node as it enabled graphql subscriptions which are used by the frontend.

Reviewer: ```suggestion
The data-node configuration supports SSL for the HTTP connection. It is highly recommended that SSL is setup on the data-node as this enables GraphQL subscriptions which are used by frontend dApps.
```

This can be setup in the following section of the data-node:
```
[Gateway]
  Level = "Info"
  Timeout = "5s"
  SubscriptionRetries = 3
  GraphQLPlaygroundEnabled = true
  [Gateway.Node]
    Port = 3007
    IP = "0.0.0.0"
  [Gateway.GraphQL]
    Port = 3008
    IP = "0.0.0.0"
    Enabled = true
    ComplexityLimit = 0
    HTTPSEnabled = false                # from here
    AutoCertDomain = ""
    CertificateFile = ""
    KeyFile = ""                        # to there
  [Gateway.REST]
    Port = 3009
    IP = "0.0.0.0"
    Enabled = true
    APMEnabled = true
```

> If you do not want to enable ssl make sure to set `HTTPSEnabled` to false or the data-node would not start properly.

Reviewer: ```suggestion
If you do not want to enable SSL, ensure `HTTPSEnabled` is set to false or the data-node will not start properly.
```

### Step 1: Stop the network

> Wait for a new checkpoint file to be produce then stop all the nodes of the network (vega, data-node and tendermint). Then backup all tendermint chain data and vega data.

Reviewer: ```suggestion
Wait for a new checkpoint file to be produced, then stop all the nodes of the network (vega, data-node and Tendermint). Once stopped backup all Tendermint chain data and vega data.
```

Save the selected checkpoint file in a safe location you will need to reuse it later.

:::info

> You can locate all your node checkpoint under: `/path/to/vega/home/vega/node/checkpoints`

Reviewer: ```suggestion
You can locate all your nodes checkpoint files under: `/path/to/vega/home/vega/node/checkpoints`
```

> You also can get a list of all path used by your node using `vega paths list`, the checkpoints folder path is `CheckpointStateHome` under this list.

Reviewer: ```suggestion
You can also get a list of all paths used by your node using `vega paths list`, the checkpoints folder path is `CheckpointStateHome` within this list.
```

:::

You can now remove all previous state of the chain by doing:
```
vega --home=/path/to/vega/home unsafe_reset_all
vega tm --home=/path/to/tendermint/home unsafe_reset_all
rm -rf /path/to/data-node/home/vega/data-node/storage/
```

:::info

> The exact path of the data-node folder to remove can be found using `vega paths list`, this path is `DataNodeStorageHome` under this list.

Reviewer: ```suggestion
The exact path of the data-node folder to remove can be found using `vega paths list`. The required path is `DataNodeStorageHome` in the list.
```

:::

### Step 2: Update the genesis file

One of the validators will now need to update the [genesis file](https://github.com/vegaprotocol/networks/blob/master/mainnet1/genesis.json) with the following information:
- the new start date of the network
- the new network ID
- the hash of the checkpoint file to be used.

This should be done via a pull request on the [networks](https://github.com/vegaprotocol/networks) repo and ideally approved by 2/3+1 of all validators.

### Step 3: Restart the network

:::note

> This is a critical part which needs to be done with all validator synchronously (this will not be needed in future updates).

Reviewer: ```suggestion
This is a critical part which needs to be done with all validators synchronously (this will not be needed in future updates).
```

:::

> All validators needs to restart their node in a synchronous way. The reason is that the network needs to synchronise it's state with the ethereum state of vega token delegation to the vega network. This is done during the bootstrapping period which happend during the N first block of the chain (this can be configured in the genesis file).

Reviewer: ```suggestion
All validators need to restart their node in a synchronous way. This is required as the network needs to synchronise its state with the Ethereum state in relations to the vega token delegation to the vega network. This is done during the bootstrapping period which happens during the first `N` blocks of the chain (`N` can be configured in the genesis file).
```

> During the bootstraping no transaction from users can be emitted apart from the transaction to submit the checkpoint. This should be done only once by one of the validators using the following command:

Reviewer: ```suggestion
During the bootstrapping no transactions from users can be emitted other than the transaction to submit the checkpoint. This should be done only once by one of the validators using the following command:
```

```
vega checkpoint restore --home=/path/to/vega/home --passphrase-file=YOUR_NODEWALLET_PASSPHRASE_FILE --checkpoint-file=/path/the/
checkpoint/file
```

Once this is done, we will need to monitor the network to make sure all delegation are recovered properly by the end of the bootstraping period

### Misc

> The ethereum event forwarder is now not required anymore, this should be removed from your infrastructure and not started. This service have been re-written and is not integrated in the vega core node.

Reviewer: ```suggestion
The ethereum event forwarder is now not required anymore, this should be removed from your infrastructure and not started. This service has been re-written and is now integrated in the vega core node.
```

Reviewer: Most comments are minor, however @jeremyletang - please confirm on this that my suggested change of the word `not` to `now` is correct. Thanks

Reviewer 2: You are correct gordon. Thanks.
